jax/_src/custom_call.py

```diff
@@ -0,0 +1,117 @@
+    backend_config: Configuration data for the custom call. If not provided,
+      this will be constructed using any keyword arguments passed when lowering.
+    result_types: The MLIR types of the results. By default, the types of the
+      output avals will be used.
+    result_shapes: Tensors that represent the result shapes, to be used when
+      the results have dynamic shapes. By default, this will be constructed
+      automatically based on the shapes of the output avals.
+    operand_layouts: A sequence of layouts (dimension orders) for each operand.
+      By default, the operands are assumed to be row-major.
+    result_layouts: A sequence of layouts (dimension orders) for each result.
+      By default, the results are assumed to be row-major.
+    lowering_args: Any other arguments to :func:`mlir.custom_call` will also be
+      passed through if provided as extra arguments to this function.
+  """
+
+  def _lowering(
+    ctx: mlir.LoweringRuleContext, *operands: ir.Value, **params: Any
+  ) -> Sequence[ir.Value | Sequence[ir.Value]]:
+    kwargs = dict(lowering_args)
+    if params and backend_config is None:
+      assert api_version >= 4
+      kwargs["backend_config"] = {k: _ir_attribute(v) for k, v in params.items()}
+    if result_types is None:
+      kwargs["result_types"] = [mlir.aval_to_ir_type(aval) for aval in ctx.avals_out]
+
+    # TODO(dfm): This is a common pattern for supporting dynamic shapes, but I
+    # haven't tested this implementation yet.
+    if result_shapes is None:
+      if any(not core.is_constant_shape(aval_out.shape) for aval_out in ctx.avals_out):
+        kwargs["result_shapes"] = [
+            mlir.shape_tensor(mlir.eval_dynamic_shape(ctx, aval_out.shape))
+            for aval_out in ctx.avals_out]
+
+    if operand_layouts is None:
+      kwargs["operand_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_in)  # pytype: disable=attribute-error
+    if result_layouts is None:
+      kwargs["result_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_out)
+
+    return mlir.custom_call(
+      call_target_name, operands=operands, api_version=api_version, **kwargs
+    ).results  # type: ignore
+
+  return _lowering
+
+
+def _default_layouts(shapes: Iterable[tuple[int]]):
+  return [list(reversed(range(len(shape)))) for shape in shapes]
+
+
+def _ir_attribute(obj: Any) -> ir.Attribute:
+  # TODO(dfm): Similar functions exist in Pallas and Mosaic GPU. Perhaps these
+  # could be consolidated into mlir or similar.
+  if isinstance(obj, str):
+    return ir.StringAttr.get(obj)
+  elif isinstance(obj, bool):
+    return ir.BoolAttr.get(obj)
+  else:
+    dtype = dtypes.canonicalize_dtype(np.result_type(obj))
+    if not (dtypes.is_python_scalar(obj) or np.isscalar(obj)):
+      raise TypeError("Only scalar attributes are supported")
+    mlir_type = mlir.dtype_to_ir_type(dtype)
+    if ir.IntegerType.isinstance(mlir_type) or ir.IndexType.isinstance(mlir_type):
```

Comment:
FYI you can do `isinstance(..., ir.IntegerType)` here and below for `ir.FloatType`. This is not true for any MLIR types, but it ints/floats have a Python type hierarchy with `ir.IntegerType` and `ir.FloatType` as base classes.

Reply:
Also, I don't think `mlir.dtype_to_ir_type` can return `ir.IndexType`?